### ISSUE TITLE
linux process "zombied" after 30 hours

### ISSUE THREAD

**Issue by atramos** [skepticoin/skepticoin#33]
I noticed a sudden decline in CPU utilization to near 0%. Did some initial investigation and noticed the following:

```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root     16491  0.0  0.6 142456 12072 ?        S    May18   0:31 /usr/bin/python3 /usr/local/bin/skepticoin-mine
root     16493 92.5  0.0      0     0 ?        Z    May18 1755:25 [skepticoin-mine] <defunct>
```
While `top` shows this:
```
top - 05:14:27 up 1 day,  7:45,  1 user,  load average: 0.00, 0.00, 0.00
Tasks: 119 total,   1 running,  77 sleeping,   0 stopped,   1 zombie
%Cpu(s):  0.3 us,  0.3 sy,  0.0 ni, 99.3 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 s
```
(note the interesting discrepancy between `ps` and `top`'s opinion of CPU usage)

The log output ends abruptly, here's the last line:
`2021-05-20 03:06:48 | uptime: 1 day, 5:35:14 |   7 hash/sec | mined: XXX SKEPTI | XX SKEPTI/h`

I realize there's not enough information here to determine root cause. Capturing in case others experiencing a similar issue can provide additional details.

**Comment by atramos** [comment]
A little more info. Note what happened to network activity while CPU utilization crashed. Could be somehow related to #31?
<img width="1621" alt="chrome_2021-05-19_22-41-25" src="https://user-images.githubusercontent.com/277808/118925415-91f13580-b8f3-11eb-857c-7a6febfb12d6.png">

**Comment by atramos** [comment]
It happened again... this time after running for only 1 hour, and it left a stack trace, around 0300GMT today:

```
miner  0 found block: 0007XXXX-0000XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
miner  0: Stopping networking thread
miner  0: Waiting for networking thread to stop
miner  0: Done; waiting for Python-exit
Process Process-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/skepticoin/networking/peer.py", line 142, in broadcast_message
    peer.send_message(message)
  File "/usr/local/lib/python3.7/site-packages/skepticoin/networking/peer.py", line 419, in send_message
    self.check_message_backlog()
  File "/usr/local/lib/python3.7/site-packages/skepticoin/networking/peer.py", line 424, in check_message_backlog
    self.start_sending()
  File "/usr/local/lib/python3.7/site-packages/skepticoin/networking/peer.py", line 427, in start_sending
    self.local_peer.selector.modify(self.sock, selectors.EVENT_READ | selectors.EVENT_WRITE, data=self)
  File "/usr/lib64/python3.7/selectors.py", line 377, in modify
    key = self._fd_to_key[self._fileobj_lookup(fileobj)]
  File "/usr/lib64/python3.7/selectors.py", line 225, in _fileobj_lookup
    return _fileobj_to_fd(fileobj)
  File "/usr/lib64/python3.7/selectors.py", line 42, in _fileobj_to_fd
    raise ValueError("Invalid file descriptor: {}".format(fd))
ValueError: Invalid file descriptor: -1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/lib64/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/site-packages/skepticoin/scripts/mine.py", line 33, in run_miner
    miner()
  File "/usr/local/lib/python3.7/site-packages/skepticoin/scripts/mine.py", line 124, in __call__
    self.handle_mined_block(block)
  File "/usr/local/lib/python3.7/site-packages/skepticoin/scripts/mine.py", line 93, in handle_mined_block
    self.thread.local_peer.network_manager.broadcast_block(block)
  File "/usr/local/lib/python3.7/site-packages/skepticoin/networking/peer.py", line 133, in broadcast_block
    self.broadcast_message(DataMessage(DATA_BLOCK, block))
  File "/usr/local/lib/python3.7/site-packages/skepticoin/networking/peer.py", line 149, in broadcast_message
    self.logger.warning(traceback.format_exc())  # be loud... this is likely a programming error.
AttributeError: 'NetworkManager' object has no attribute 'logger'
```

**Comment by lk16** [comment]
> AttributeError: 'NetworkManager' object has no attribute 'logger'

This should be `self.local_peer.logger` instead of `self.logger`.
I have a fix for this in my open PR #16, [here](https://github.com/skepticoin/skepticoin/pull/16/files#diff-ed3401667bfac5d3097a789b8c08fd87b8d3c4d2d2d53f18870cda39bcf9c355L149-L150).

**Comment by sashimi-houdini** [comment]
> This should be `self.local_peer.logger` instead of `self.logger`.
> I have a fix for this in my open PR #16, [here](https://github.com/skepticoin/skepticoin/pull/16/files#diff-ed3401667bfac5d3097a789b8c08fd87b8d3c4d2d2d53f18870cda39bcf9c355L149-L150).

This is true, and the fix has made it into `master` already. That takes care of the part of the stacktrace below the line "During handling of the above exception", but not the part above it.
